bin/sozo/src/commands/auth.rs

```diff
@@ -15,17 +19,131 @@ pub struct AuthArgs {
+            .map_err(|_| anyhow::anyhow!("Invalid contract address: {}", contract_part))?;
+
+        Ok(ModelContract { model, contract })
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum ResourceType {
+    Contract(String),
+    Model(FieldElement),
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct OwnerResource {
+    pub owner: FieldElement,
+    pub resource: ResourceType,
+}
+
+impl FromStr for OwnerResource {
+    type Err = anyhow::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let parts: Vec<&str> = s.split(',').collect();
+
+        let (owner_part, resource_part) = match parts.as_slice() {
+            [owner, resource] => (*owner, *resource),
+            _ => anyhow::bail!(
+                "Owner and resource are expected to be comma separated: `sozo auth owner \
+                 0x1234,resource_type:resource_name`"
+            ),
+        };
+
+        let owner = FieldElement::from_hex_be(owner_part)
+            .map_err(|_| anyhow::anyhow!("Invalid owner address: {}", owner_part))?;
+
+        let resource_parts: Vec<&str> = resource_part.split(':').collect();
+        let resource = match resource_parts.as_slice() {
+            ["contract", name] => ResourceType::Contract(name.to_string()),
+            ["model", name] => {
+                let model = cairo_utils::str_to_felt(name)
+                    .map_err(|_| anyhow::anyhow!("Invalid model name: {}", name))?;
+                ResourceType::Model(model)
+            }
+            _ => anyhow::bail!(
+                "Resource is expected to be in the format `resource_type:resource_name`: `sozo \
+                 auth owner 0x1234,resource_type:resource_name`"
+            ),
+        };
+
+        Ok(OwnerResource { owner, resource })
+    }
+}
+
 #[derive(Debug, Subcommand)]
-pub enum AuthCommand {
-    #[command(about = "Auth a system with the given calldata.")]
+pub enum AuthKind {
+    #[command(about = "Grant a contract permission to write to a model.")]
     Writer {
         #[arg(num_args = 1..)]
         #[arg(required = true)]
         #[arg(value_name = "model,contract_address")]
         #[arg(help = "A list of models and contract address to grant write access to. Comma \
                       separated values to indicate model name and contract address e.g. \
                       model_name,0x1234 model_name,0x1111 ")]
-        models_contracts: Vec<String>,
+        models_contracts: Vec<ModelContract>,
+    },
+    #[command(about = "Grant ownership of a resource.")]
+    Owner {
+        #[arg(num_args = 1..)]
+        #[arg(required = true)]
+        #[arg(value_name = "owner,resource")]
+        #[arg(help = "A list of owners and resources to grant ownership to. Comma separated \
+                      values to indicate owner address and resouce e.g. \
+                      0x1234,contract:contract_name 0x1111,contract:contract_address, \
+                      0xbeef,model:model_name")]
```

Comment:
It feels more natural this way. We may then change `sozo auth writer` to also have the contract address first and the model name after.
```
sozo auth grant writer 0x1234,model_name
sozo auth grant owner 0x1234,model:model_name
```
WDYT?

Reply:
i was thinking about it and i think actually your original ordering makes more sense, since the second value is being applied to the first, so i'll update to follow yours 😆